### PR TITLE
RESTWS-658 : Person web service allow checking if person has a Patien…

### PR DESCRIPTION
**Pull request by Ruhanga**
## Description of what I changed

Introduced `Person` object in the `PersonResource1_8` class to help evaluate whether person in question is patient.
Introduced a method to specify `patient` resource/endpoint in the `description` 's links for the `person`  in question.

## Issue I worked on

see https://issues.openmrs.org/browse/RESTWS-658

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**

- [x] My IDE is configured to follow the [**code style**]

- [x] I have **added tests** to cover my changes. 

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

- [x] All new and existing **tests passed**.

- [x] My pull request is **based on the latest changes** of the master branch.

### CODE REVIEW

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java [hunk 1]
```diff
@@ -552,4 +556,19 @@ public SimpleObject getAuditInfo(Person person) throws Exception {
...
+	
+	/**
+	 * Adds the link to patient resource in this person description if person is patient
+	 * 
+	 * @param description whose link set is to be modified
+	 * @return audit resource description
+	 */
+	private DelegatingResourceDescription ifPatientAddLink(DelegatingResourceDescription description) {
+		if (this.person != null && person.isPatient()) {
+			Patient patient = Context.getPatientService().getPatient(this.person.getId());
```
Comment (samuelmale) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java]: What happens if the returned Patient is null?

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java [hunk 2]
```diff
@@ -391,7 +394,8 @@ public static PersonAddress getPreferredAddress(Person instance) {
 	 */
 	@Override
 	public Person getByUniqueId(String uuid) {
-		return Context.getPersonService().getPersonByUuid(uuid);
+		this.person = Context.getPersonService().getPersonByUuid(uuid);
```
Comment (samuelmale) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java]: Why did you change this?

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java [hunk 3]
```diff
@@ -54,6 +55,8 @@
 //order must be greater than that for PatientResource(order=0) RESTWS-273
 public class PersonResource1_8 extends DataDelegatingCrudResource<Person> {
 	
+	private Person person = null;
```
Comment (samuelmale) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java]: Any inevitable reason of making this a class level variable?

Comment (Ruhanga) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java]: I need a way to reference the `person` in question  and in this context can only be initialized by the nearest defined method `getByUniqueId(String uuid)` in this class as I have refactored. @samuelmale , could there be an alternative to exposing this `person` object to this resource class without refactoring any super class to satisfy this need?

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java (hunk 2 above)
Comment (Ruhanga) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java]: I have presented my thought on this in the comment on the `Person` class level variable.

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java [hunk 4]
```diff
@@ -552,4 +556,19 @@ public SimpleObject getAuditInfo(Person person) throws Exception {
 		ret.put("dateCreated", ConversionUtil.convertToRepresentation(person.getPersonDateCreated(), Representation.DEFAULT));
 		return ret;
 	}
+	
+	/**
+	 * Adds the link to patient resource in this person description if person is patient
```
[the remark this answers is not in the document]
Comment (Ruhanga) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java]: Sure @samuelmale , thank you. I will refactor to consider this.

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java (hunk 1 above)
Comment (Ruhanga) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java]: This is a redundant check though I included it. I just felt it safe to include it even when it is addresed by [this line ](https://github.com/openmrs/openmrs-module-webservices.rest/blob/master/omod-common/src/main/java/org/openmrs/module/webservices/rest/web/resource/impl/DelegatingCrudResource.java#L52). I am now concluding to remove it (this.person != null). Thank you @samuelmale.

Comment (samuelmale) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java]: I think you did get me. I'm not talking about the if statement check but rather the returned `patient`. What happens if the returned value is null?

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java (hunk 3 above)
Comment (samuelmale) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java]: I believe the original `getByUniqueId` method was enough to do what you want.
What happens if you remove this from the class level and edit yr changes. Does the test fail?

Comment (Ruhanga) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java]: For me to call this method within this resource class, I need to supply the `uuid` argument which is not anywhere referenced atleast in current class context. This, on the other hand is handled for me or called by DelegatingCrudResource-supper class [here](https://github.com/openmrs/openmrs-module-webservices.rest/blob/master/omod-common/src/main/java/org/openmrs/module/webservices/rest/web/resource/impl/DelegatingCrudResource.java#L51) and so with the changes I have made, `person` object is initialized at Run time when the method is called in the line I have referred to.

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java (hunk 1 above)
Comment (Ruhanga) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java]: Ohh, good catch @samuelmale. I just didn't expect the related to patient to be null though. In the case this happens, a null pointer exception would be thrown from [this line](https://github.com/openmrs/openmrs-module-webservices.rest/pull/379/commits/f417178211fff0ab8109afe2a7300791a40b15a1#diff-060b1f85f075bfef61fde7654399e5c5R569)

Comment (samuelmale) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java]: Do you wan't to prevent the possibility of a NPE?

File: omod-1.9/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_9/PersonController1_9Test.java [hunk 5]
```diff
@@ -82,6 +82,23 @@ public void shouldGetAPersonByUuid() throws Exception {
...
+		String recievedLink = null;
+		for (Object link : links.toArray()) {
+			if ("patient".equals((String) PropertyUtils.getProperty(link, "rel"))) {
+				recievedLink = (String) PropertyUtils.getProperty(link, "uri");
+			}
+		}
+		assertNotNull(recievedLink);
+		assertEquals(expectedLink, recievedLink);
+	}
+
```
Comment (samuelmale) [inline omod-1.9/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_9/PersonController1_9Test.java]: Any reason for putting this test in the omod-1.9 instead of omod-1.8?

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java [hunk 6]
```diff
@@ -552,4 +556,19 @@ public SimpleObject getAuditInfo(Person person) throws Exception {
...
+	 * 
+	 * @param description whose link set is to be modified
+	 * @return audit resource description
+	 */
+	private DelegatingResourceDescription ifPatientAddLink(DelegatingResourceDescription description) {
```
Comment (samuelmale) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java]: Can we improve on the method name to reflect more meaning? eg. `addLinkIfPatient`, `addLinkToPatientIfRequired` etc. sounds better.

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java (hunk 3 above)
Comment (samuelmale) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java]: Makes sense to me.

File: omod-1.9/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_9/PersonController1_9Test.java [hunk 7]
```diff
@@ -82,6 +82,23 @@ public void shouldGetAPersonByUuid() throws Exception {
...
+		List<Object> links = (List<Object>) result.get("links");
+		String expectedLink = RestConstants.URI_PREFIX + "v1/patient/" + RestTestConstants1_8.PATIENT_UUID;
+		String recievedLink = null;
+		for (Object link : links.toArray()) {
+			if ("patient".equals((String) PropertyUtils.getProperty(link, "rel"))) {
+				recievedLink = (String) PropertyUtils.getProperty(link, "uri");
+			}
+		}
+		assertNotNull(recievedLink);
+		assertEquals(expectedLink, recievedLink);
```
Comment (samuelmale) [inline omod-1.9/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_9/PersonController1_9Test.java]: Can you also test for cases where Person is not a Patient?

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java (hunk 1 above)
Comment (Ruhanga) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java]: The exception is adequate enough to describe the problem in the case the unexpected happens. @samuelmale , what other other alternative  would work best?

File: omod-1.9/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_9/PersonController1_9Test.java (hunk 5 above)
Comment (Ruhanga) [reply omod-1.9/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_9/PersonController1_9Test.java]: All other tests on other functionalities are placed here. Same convention is followed by other test suites to other resources. Is there any reason not follow this?

File: omod-1.9/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_9/PersonController1_9Test.java (hunk 7 above)
Comment (Ruhanga) [reply omod-1.9/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_9/PersonController1_9Test.java]: Testing for possibly of this has 0% code coverage and hence I think is not necessary. The `If` clause I used was without doubt, explicit in only branching and executing block of of code in it that needs to be covered while unit testing otherwise nothing is executed, so no need to test nothing-executing.

**Comment by samuelmale** [comment]
@dkayiwa do you have any objections?

**Comment by dkayiwa** [comment]
On the JIRA ticket, can you share a screenshot with the effects of your changes?

**Comment by Ruhanga** [comment]
Alright @dkayiwa , let attach the images to JIRA ticket.

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java (hunk 3 above)
Comment (dkayiwa) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java]: These resources should be stateless. So can we remove the class level variable?

Comment (Ruhanga) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java]: Enforcing this constraint for this particular use-case implies making a huge refactor on the `getRepresentationDescription(Representation rep)` method to `getRepresentationDescription(Representation rep Person person)` or it's more generic form where all the implementing classes have to be refactored as well, which is not work related to this ticket being addressed. @dkayiwa , could there be a better way of passing reference, `delegate`/`person` to this resource without affecting the rest of the resources, or I create another ticket to refactor `getRepresentationDescription(Representation rep)` for all implementing resources?

Comment (haripriya999) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java]: Is it any good in assigning the value of person to null ?

Comment (dkayiwa) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PersonResource1_8.java]: We are gonna have to look for another solution because resources should be stateless.

**Comment by dkayiwa** [comment]
Closing this because we are gonna have to look for another solution since resources should be stateless.
